Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesWindow.xaml.cs

```diff
@@ -18,14 +20,88 @@ public void ClearEntries()
         }
 
+        public IEnumerable<Collapsible> GetAllCollapsibleBoxes()
+        {
+            return EntryContainer.Children.OfType<Collapsible>();
+        }
+
+        public Collapsible? GetCollapsibleById(string id)
+        {
+            return EntryContainer.Children
+                .OfType<Collapsible>()
+                .FirstOrDefault(c => c.Name == id);
+        }
+
+        public Dictionary<int, bool> SaveCollapsibleBoxesStates()
+        {
+            var collapsibleStates = new Dictionary<int, bool>();
+            foreach (var collapsible in GetAllCollapsibleBoxes())
+            {
+                if (int.TryParse(collapsible.Name, out var collapsibleId))
+                {
+                    collapsibleStates[collapsibleId] = collapsible.BodyVisible;
+                }
+            }
+            return collapsibleStates;
+        }
+
+        public void RestoreCollapsibleBoxesStates(Dictionary<int, bool> collapsibleStates)
+        {
+            foreach (var collapsible in GetAllCollapsibleBoxes())
+            {
+                if (int.TryParse(collapsible.Name, out var collapsibleId) && collapsibleStates.TryGetValue(collapsibleId, out var isOpen))
+                {
+                    collapsible.BodyVisible = isOpen;
+                }
+            }
+        }
+
         public void AddEntry(string name, string description, bool hasAccess, FormattedMessage? reason, IEnumerable<GhostRoleInfo> roles, SpriteSystem spriteSystem)
         {
             NoRolesMessage.Visible = false;
 
-            var entry = new GhostRolesEntry(name, description, hasAccess, reason, roles, spriteSystem);
-            entry.OnRoleSelected += OnRoleRequestButtonClicked;
-            entry.OnRoleFollow += OnRoleFollow;
-            EntryContainer.AddChild(entry);
+            var ghostRoleInfos = roles.ToList();
+            var rolesCount = ghostRoleInfos.Count;
+
+            var info = new GhostRoleInfoBox(name, description);
+            var buttons = new GhostRoleButtonsBox(hasAccess, reason, ghostRoleInfos, spriteSystem);
+            buttons.OnRoleSelected += OnRoleRequestButtonClicked;
+            buttons.OnRoleFollow += OnRoleFollow;
+
+            EntryContainer.AddChild(info);
+
+            if (rolesCount > 1)
+            {
+                var buttonHeading = new CollapsibleHeading(Loc.GetString("ghost-roles-window-available-button", ("rolesCount", rolesCount)));
+
+                buttonHeading.AddStyleClass(ContainerButton.StyleClassButton);
+                buttonHeading.HorizontalAlignment = HAlignment.Stretch;
+                buttonHeading.Label.HorizontalAlignment = HAlignment.Center;
+                buttonHeading.Label.HorizontalExpand = true;
+                buttonHeading.HorizontalExpand = true;
+
+                var body = new CollapsibleBody
+                {
+                    Margin = new Thickness(0, 10, 0, 0),
+                };
+
+                var uniqueId = name.GetHashCode();
```

Comment:
Hash codes are not guaranteed to be unique IDs. You need to key this data the same way they're grouped in the EUI code.